File: data_structures/arrays/005MoveNegatives/negative.cpp
#include <bits/stdc++.h>
using namespace std;

void move_negatives(int array[], int len) {
  int j = 0;
  for (int i = 0; i < len; i++) {
    if (array[i] < 0) {
      swap(array[i], array[j]);
      j++;
    }
  }
}

void move_using_two_pointers(int array[], int len) {
  int left = 0;
  int right = len - 1;

  while (left <= right) {
    if (array[left] < 0 && array[right] < 0) {
      left++;
    } else if (array[left] > 0 && array[right] < 0) {
      swap(array[left], array[right]);
      left++;
      right--;
    } else if (array[left] > 0 && array[right] > 0) {
      right--;
    } else {
      left++;
      right--;
    }
  }
}

int main() {
  // code
  int array[] = {1, 24, -15, -1, 5125, 12, -1241, -4124, -124151, -12};
  int len = sizeof(array) / sizeof(array[0]);
  // move_negatives(array, len);
  move_using_two_pointers(array, len);
  cout << "The Modified Array Is:\n";
  for (int i = 0; i < len; i++) {
    cout << array[i] << " ";
  }
  cout << "\n";
}

File: data_structures/arrays/005MoveNegatives/negative.py
def move_negatives(array: list) -> None:
    size = len(array)
    j = 0
    for i in range(size):
        if array[i] < 0 :
            array[j], array[i] = array[i], array[j]
            j += 1

def move_using_two_pointers(array: list, size: int) -> None:
    left = 0
    right = size - 1

    while left <= right:
        if (array[left] < 0 and array[right] < 0):
            left += 1
        elif (array[left] > 0 and array[right] < 0):
            array[left], array[right] = array[right], array[left]
            left += 1
            right -= 1
        elif (array[left] > 0 and array[right] > 0):
            right -= 1
        else:
            left += 1
            right -= 1
            
def main():
    array = [1,1,24,51,51,215,125,6451,-12,-89]
    move_negatives(array)
    print("Modified Array Is:")
    print(array)

if __name__ == "__main__":
    main()